src/Common/src/System/IO/PathInternal.Windows.cs

```diff
@@ -19,19 +42,67 @@ internal static bool HasIllegalCharacters(string path, bool checkAdditional = fa
 
             Contract.Requires(path != null);
 
-            foreach (char c in path)
+            // Question mark is a normal part of extended path syntax (\\?\)
+            int startIndex = path.StartsWith(LongPathPrefix, StringComparison.Ordinal) ? LongPathPrefix.Length : 0;
+            return path.IndexOfAny(checkAdditional ? InvalidPathCharsWithAdditionalChecks : InvalidPathChars, startIndex) >= 0;
+        }
+
+        /// <summary>
+        /// Gets the length of the root of the path (drive, share, etc.).
+        /// </summary>
+        internal static int GetRootLength(string path)
+        {
+            CheckInvalidPathChars(path);
+
+            int i = 0;
+            int length = path.Length;
+            int volumeSeparatorLength = 2;  // Length to the colon "C:"
+            int uncRootLength = 2;          // Length to the start of the server name "\\"
+
+            bool extendedSyntax = path.StartsWith(LongPathPrefix, StringComparison.Ordinal);
+            bool extendedUncSyntax = extendedSyntax && path.StartsWith(UNCLongPathPrefix, StringComparison.Ordinal);
```

Comment:
Yeah, I was thinking the same thing. I erred towards clarity for the initial implementations. My thought was it would be best to go back and fine tune for perf later based on sets of various orders of magnitude and types (100/1000 characters, relative/non-relative, 100/10K/100K paths etc.).